x/stake/keeper/delegation.go

```diff
@@ -0,0 +1,230 @@
+package keeper
+
+import (
+	"bytes"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/stake/types"
+)
+
+// load a delegation
+func (k Keeper) GetDelegation(ctx sdk.Context,
+	delegatorAddr, validatorAddr sdk.Address) (delegation types.Delegation, found bool) {
+
+	store := ctx.KVStore(k.storeKey)
+	delegatorBytes := store.Get(GetDelegationKey(delegatorAddr, validatorAddr, k.cdc))
+	if delegatorBytes == nil {
+		return delegation, false
+	}
+
+	k.cdc.MustUnmarshalBinary(delegatorBytes, &delegation)
+	return delegation, true
+}
+
+// load all delegations used during genesis dump
+func (k Keeper) GetAllDelegations(ctx sdk.Context) (delegations []types.Delegation) {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, DelegationKey)
+
+	i := 0
+	for ; ; i++ {
+		if !iterator.Valid() {
+			iterator.Close()
+			break
+		}
+		bondBytes := iterator.Value()
+		var delegation types.Delegation
+		k.cdc.MustUnmarshalBinary(bondBytes, &delegation)
+		delegations = append(delegations, delegation)
+		iterator.Next()
+	}
+	return delegations[:i] // trim
+}
+
+// load all delegations for a delegator
+func (k Keeper) GetDelegations(ctx sdk.Context, delegator sdk.Address,
+	maxRetrieve int16) (delegations []types.Delegation) {
+
+	store := ctx.KVStore(k.storeKey)
+	delegatorPrefixKey := GetDelegationsKey(delegator, k.cdc)
+	iterator := sdk.KVStorePrefixIterator(store, delegatorPrefixKey) //smallest to largest
+
+	delegations = make([]types.Delegation, maxRetrieve)
+	i := 0
+	for ; ; i++ {
+		if !iterator.Valid() || i > int(maxRetrieve-1) {
+			iterator.Close()
+			break
+		}
+		bondBytes := iterator.Value()
+		var delegation types.Delegation
+		k.cdc.MustUnmarshalBinary(bondBytes, &delegation)
+		delegations[i] = delegation
+		iterator.Next()
+	}
+	return delegations[:i] // trim
+}
+
+// set the delegation
+func (k Keeper) SetDelegation(ctx sdk.Context, delegation types.Delegation) {
+	store := ctx.KVStore(k.storeKey)
+	b := k.cdc.MustMarshalBinary(delegation)
+	store.Set(GetDelegationKey(delegation.DelegatorAddr, delegation.ValidatorAddr, k.cdc), b)
+}
+
+// remove the delegation
+func (k Keeper) RemoveDelegation(ctx sdk.Context, delegation types.Delegation) {
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(GetDelegationKey(delegation.DelegatorAddr, delegation.ValidatorAddr, k.cdc))
+}
+
+// common functionality between handlers
+func (k Keeper) Delegate(ctx sdk.Context, delegatorAddr sdk.Address, bondAmt sdk.Coin,
```

Comment:
I think functions like this, which write some state updates (subtracting coins) to the store but *don't* write others (validator / delegator changes) are a bit confusing, easy to call them in the wrong way. Do you think we could structure staking so that functions *either* write all their updates to the store or write nothing and return the updated parameters?

Reply:
Great point, I modified to now make and save/update all the changed objects in the state, seemed like the more natural grouping of code as opposed to not modifying anything. The "not-modifying-anything" functions are great for non-keeper types. aka there are lots of objects in the `types` directory which return modified objects, which are then expected to be saved later.

Oh I also modified the other function in this file (now `func Unbond`) to act in the same way, it updates and saved the objects instead of passing them back